egraph.rkt

```diff
@@ -133,80 +172,108 @@
-          (values l1 l2 old-vars2)
-          (values l2 l1 old-vars1)))
-
-    ;; Get the expressions which mention the follower so we can see if
-    ;; their new form causes new merges.
-    (define iexprs (hash-ref leader->iexprs follower))
-
-    ;; Once we've merged these enodes, other ones might have become
-    ;; equivalent. For example, if we had an enode which had the
-    ;; variation (+ x 1), and an enode which had the variation (+ y
-    ;; 1), and we merged x and y, then we know that these two enodes
-    ;; are equivalent, and should be merged.
-    (define to-merge
-      (for/list ([iexpr (in-mutable-set iexprs)])
-        (define replaced-iexpr (update-en-expr iexpr))
-        (define other-parent (hash-ref expr->parent replaced-iexpr #f))
-        (and other-parent
-             (cons other-parent (hash-ref expr->parent iexpr)))))
-
-    ;; Now that we have extracted all the information we need from the
-    ;; egraph maps in their current state, we are ready to update
-    ;; them. We need to know which one is the old leader, and which is
-    ;; the new to easily do this, so we branch on which one is eq? to
-    ;; merged-en.
-    (update-leader! eg follower-old-vars follower leader)
-
-    ;; Now the state is consistent for this merge, so we can tackle
-    ;; the other merges.
-    (for ([node-pair (in-list to-merge)] #:when node-pair)
-      (merge-egraph-nodes! eg (car node-pair) (cdr node-pair)))
-
-    ;; The other merges can have caused new things to merge with our
-    ;; merged-en from before (due to loops in the egraph), so we turn
-    ;; this into a leader before finally returning it.
-    (pack-leader merged-en)]))
-
-(define (update-en-expr expr)
-  (if (list? expr)
-      (for/list ([sub (in-list expr)])
-        (if (enode? sub) (pack-leader sub) sub))
-      expr))
+  (define res
+    (cond
+      [(eq? l1 l2)
+       ;; If the leaders are the same, then these nodes are already part
+       ;; of the same pack. However, this call usually means that two
+       ;; vars of this leader were found equivalent through another
+       ;; merge, so we want to update the vars to remove the redundancy.
+       (unless (eq? en1 en2)
+         (dedup-vars! l1))]
+      [else
+       ;; Hold on to these vars as they won't be the same after the
+       ;; merge, but we don't yet know which one we need.
+       (define old-vars1 (enode-vars l1))
+       (define old-vars2 (enode-vars l2))
+
+       ;; Merge the node packs
+       (define merged-en (enode-merge! l1 l2))
+
+       ;; Now that we know which one became leader, we can bind these.
+       (define-values (leader follower follower-old-vars)
+         (if (eq? l1 merged-en)
+             (values l1 l2 old-vars2)
+             (values l2 l1 old-vars1)))
+
+       (unless rebuilding-enabled?
+         ;; update expr->parent, discovering parent nodes to merge
+         (define to-merge
+           (update-expr->parent! eg follower-old-vars follower leader))
```

Comment:
**Bug 3**
Upwards merging was looking at collisions before doing hashcons surgery. This led to an edge case involving two enodes that are both out of date. Now it discovers merges while doing surgery in this call.